In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
def var_historic(r, level=1):
    """
    Takes in a series of returns (r), and the percentage level
(level)
    Returns the historic Value at Risk at a specified level
    i.e. returns the number such that "level" percent of the returns
    fall below that number, and the (100-level) percent are above
    """
    if isinstance(r, pd.DataFrame):
        return r.aggregate(var_historic, level=level)
    elif isinstance(r, pd.Series):
        return -np.percentile(r, level)
    else:
        raise TypeError("Expected r to be a Series or DataFrame")

# Pregunta 1 - 2

In [196]:
df = pd.read_excel("var1_2.xlsx")
df = df.sort_values(by="val",ascending = False).reset_index()
df.iloc[10,2]

43895.21521742776

## Pregunta 1

In [21]:
#var histórico
var_historic(df["val"], level=99)

-43897.45709514318

### Respuesta
R/ 43895

## Pregunta 2

In [39]:
def var_parametric(df, level):
    var = np.percentile(df, level)
    return(var)

In [40]:
def c_var(df, level):
    value_at_risk = var_parametric(df, level)
    cvar = df[df <= value_at_risk].mean()
    return(cvar)

In [198]:
df_cvar = df.iloc[0:10,2]
df_cvar.mean()

48246.101495835544

### Respuesta
R/ 48246

# Pregunta 3

In [199]:
df3 = pd.read_excel("var_3.xlsx")
df3.drop(columns = "id", inplace = True)
df3.head()

Simulación 1  Simulación 2  Simulación 3  Simulación 4  Simulación 5
0  48835.522211  33764.139378  53763.127375  65114.423931  48902.652385
1  57411.454449  64582.123575  25662.026770  40053.897055  46849.128198
2  42227.836115  44432.550153  44118.271581  37462.771696  59485.864254
3  67762.080368  39562.468729  30236.228537  44532.352384  41677.000504
4  63698.821328  39110.547853  39921.080138  51967.998906  44923.587496

In [219]:
var_list = []
cvar_list = []
largo = len(df3)
var_level = 0.01
idx = int(largo * var_level)
for col in df3.columns:
    df_temp = df3[col]
    df_temp_sorted = df_temp.sort_values(ascending = False).reset_index()
    var = df_temp_sorted.iloc[idx,1]
    c_var = df_temp_sorted.iloc[0:idx,1].mean() 
    var_list.append(var)
    cvar_list.append(c_var)
    

In [220]:
print("Std VaR", statistics.stdev(var_list))
print("Std CVaR", statistics.stdev(cvar_list))

Std VaR 1380.6308063562174
Std CVaR 1608.4567475244214


### Respuesta
R/ Var 1380 y TVar 1608

# Pregunta 4

Suponga que los siniestros que puede tener una compañía en un año con la cartera de seguros que tiene suscrita, siguen una distribución normal con media USD 10 millones y desviación estándar USD 2 millones. Si la compañía tiene en reservas la media de las pérdidas más un margen de riesgo del 5% adicional ¿Cuánto capital deben disponer los accionistas para cumplir con un nivel de solvencia basado en un VaR 99%? (Suponga que el valor de la reserva y el margen de riesgo están respaldados por activos que se obtuvieron de las primas suscritas)

In [146]:
mean = 10
std = 2
reserves = mean * 1.05
level = 0.99
var99 = norm.ppf(level, loc=mean, scale=std)
var99

14.652695748081682

In [151]:
(var99 - reserves)*1000

4152.695748081682

### Respuesta
4152

# Pregunta 5

# Pregunta 6

El horizonte temporal contempla todos los flujos de caja hasta su run-off pero solo contempla eventos que se puedan materializar durante un año

# Pregunta 7

Verdadero

# Pregunta 8

La solvencia tiene como principal objetivo dar confianza y proteger a los consumidores del sistema financiero

Los modelos internos de solvencia tienen la ventaja de reflejar de una manera más precisa los riesgos de una organización